### PR TITLE
Implement publics for Halo2 backend

### LINKED ISSUE #820: Public Inputs / Outputs
We should implement public inputs / outputs.

Actually, I'm not sure how exactly public inputs work. There might be multiple ways to do this.

It seems like in [Polygon PIL](https://wiki.polygon.technology/docs/zkevm/PIL/public-values/), a public value is basically a variable that can be part of an identity (e.g.: `ISLAST * (a - :result) = 0;`). I assume that the way this works is just that the verifier plugs it into the equation (s)he is checking. As a consequence, for each constrained cell, we have to add both a selector (in the example `ISLAST`) and an identity.

In Halo2, as far as I remember, there is a special `Instance` column type and a `layouter.constrain_instance()` function, which essentially creates a copy constraint between the specified cell and a cell in the instance column. The prover then opens the commitment of the instance column at the relevant indices, and the verification works like any other copy constraint.

These seem to be different approaches, and I'm not to sure how to reconcile those. But I might also be misunderstanding them.

### PR DESCRIPTION
**Pull request by georgwiese**
Fixes #820 
Step towards #814 

This PR implements public outputs for the Halo2 backend (for eStark, they were already supported).

### CODE REVIEW

File: halo2/src/prover.rs [hunk 1]
```diff
@@ -119,7 +118,12 @@ impl Halo2Prover {
         };
 
         log::info!("Generating circuit for app snark...");
-        let circuit_app = analyzed_to_circuit(pil, fixed, witness);
+        let (circuit_app, publics) = analyzed_to_circuit(pil, fixed, witness);
+
+        assert_eq!(publics.len(), 1);
+        if !publics[0].is_empty() {
+            unimplemented!("Public inputs are not supported yet");
```
Comment (georgwiese) [inline halo2/src/prover.rs:125]: I'm not that deep into how the aggregation circuit works and I'm not sure if we have proper tests for it. So I decided to do that in another PR.

Comment (leonardoalt) [reply halo2/src/prover.rs:125]: yea we don't really have structured tests for it, let's do it later. It's tricky to have tests for it because the aggregation circuit is pretty big so we can't run it in our current CI

File: halo2/src/circuit_builder.rs [hunk 2]
```diff
@@ -166,6 +170,38 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
...
         }
     }
+
+    assert!(
+        analyzed.public_declarations.len() <= num_rows,
+        "More publics than rows!"
+    );
+
+    // Enforce publics by copy-constraining to cells in the instance column.
+    let mut copys = vec![];
```
Comment (leonardoalt) [inline halo2/src/circuit_builder.rs:180]:
```suggestion
    let mut copies = vec![];
```

File: halo2/src/circuit_builder.rs [hunk 3]
```diff
@@ -166,6 +170,38 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
...
+        let witness_col = cd.col(&witness_name);
+        let witness_offset = public_declaration.index as usize;
+        let public_col = cd.public_column;
+        let public_offset = copys.len();
+
+        // Add copy constraint
+        copys.push(CopyC {
+            columns: (public_col, witness_col),
+            // We could also create one copy constraint per column pair wih several offsets.
+            // I don't thin there is a difference though...
```
Comment (leonardoalt) [inline halo2/src/circuit_builder.rs:194]:
```suggestion
            // I don't think there is a difference though...
```

File: halo2/src/circuit_builder.rs [hunk 4]
```diff
@@ -166,6 +170,38 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
...
+            offsets: vec![(public_offset, witness_offset)],
+        });
+
+        // Evaluate the given cell and add it to the publics.
+        let value = cd.eval_witness(&witness_name, witness_offset).to_bytes_le();
+        let mut buffer = [0u8; 64];
+        buffer[..value.len()].copy_from_slice(&value);
+        let value = Fr::from_uniform_bytes(&buffer);
```
Comment (leonardoalt) [inline halo2/src/circuit_builder.rs:202]: What's the type of `value`? Is this the easiest way to do the type conversion? Don't we have a routine already somewhere here?

File: halo2/src/circuit_builder.rs [hunk 5]
```diff
@@ -166,6 +170,38 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
             _ => unimplemented!(),
         }
     }
+
+    assert!(
+        analyzed.public_declarations.len() <= num_rows,
+        "More publics than rows!"
+    );
+
+    // Enforce publics by copy-constraining to cells in the instance column.
```
Comment (leonardoalt) [inline halo2/src/circuit_builder.rs:179]: Can you add a comment with a small example diagram here of what exactly is happening?

File: halo2/src/circuit_builder.rs (hunk 2 above)
Comment (georgwiese) [reply halo2/src/circuit_builder.rs:180]: True! But funny that it's called `copys` in Plaf, so now we pass `copys: copies` :D

File: halo2/src/circuit_builder.rs (hunk 5 above)
Comment (georgwiese) [reply halo2/src/circuit_builder.rs:179]: Done!

File: halo2/src/circuit_builder.rs (hunk 4 above)
Comment (georgwiese) [reply halo2/src/circuit_builder.rs:205]: I didn't find any :/ It converts from `Bn254Field` to `halo2_curves::bn256::Fr` by going to little endian bytes and back.

Comment (leonardoalt) [reply halo2/src/circuit_builder.rs:205]: But shouldnt we already have that somewhere here for the witness? I'll take a look in a sec

Comment (leonardoalt) [reply halo2/src/circuit_builder.rs:205]: Ah the one I had in mind is from constants and not FieldElement

File: halo2/src/circuit_builder.rs (hunk 5 above)
Comment (leonardoalt) [reply halo2/src/circuit_builder.rs:179]: where is it?

Comment (georgwiese) [reply halo2/src/circuit_builder.rs:179]: Oh, it got lost somehow? 👀 I just wrote it again, now it's there.

File: halo2/src/circuit_builder.rs [hunk 6]
```diff
@@ -166,6 +170,54 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
...
+    // - public out1 = A(2);
+    // - public out2 = B(0);
+    // This would lead to the corresponding values being copied into the public
+    // column as follows:
+    // | Row | A   | B   | public |
+    // |-----|-----|-----|--------|
+    // |  0  |  0  | *5* |  *2*   |
+    // |  1  |  1  |  6  |  *5*   |
+    // |  2  | *2* |  7  |        |
+    // |  3  |  4  |  8  |        |
```
Comment (leonardoalt) [inline halo2/src/circuit_builder.rs:190]: So the syntactic id of the public column is its index in the `public` column, right?

**Review by leonardoalt — APPROVED** [review]

File: halo2/src/circuit_builder.rs (hunk 6 above)
Comment (georgwiese) [reply halo2/src/circuit_builder.rs:190]: The index in the public columm is the index of the public declaration in the PIL file.
